doc/StaticWorkflow.md

```diff
@@ -0,0 +1,266 @@
+    3. The Edge Proxy redirects this request to the proper endpoint on the
+       Preferences Server.
+2. The Preferences Server sends an [authentication request](https://developers.google.com/identity/protocols/oauth2/openid-connect#sendauthrequest)
+   to Google.
+   1. the authentication request includes the following as query parameters:
+      * `client_id` identifying the Preferences Server to Google
+      * `response_type` which is usually `code`
+      * `scope` which would likely be `openid email`
+      * `redirect_uri` which is the endpoint on the Preferences Server that
+           will receive the response from Google. It must match the authorized
+           redirect URI that was pre-registered with Google.
+      * `state`, the anti-forgery unique session token
+   2. Login and consent:  The Resource Owner may be presented with a login
+      screen by Google in their domain, and asked to consent the requested
+      scope.
+      * If the user is already logged into Google, they will be presented
+        with the consent dialog.
+   3. The Resource Owner authenticates with Google and grants consent
+      * If the user has previously provided consent, Google will present no
+        dialog, but will retrieve the scope(s) of information for which the user
+        has consented.
+   4. [Authorization code and anti-forgery](https://developers.google.com/identity/protocols/oauth2/openid-connect#confirmxsrftoken):
+      Google responds to the Preferences Server at the `redirect_uri` including:
+      * `state` anti-forgery token from step 2i
+      * `code` the authentication code provided by Google
+      * `scope` the scopes for which the user consented at step 2iii.
+   5. The Preferences Server confirms that the `state` (anti-forgery token) is
+      valid by checking that it matches the value it sent to Google in step 2i.
+   6. [Exchange Authorization Code](https://developers.google.com/identity/protocols/oauth2/openid-connect#exchangecode):
+      The Preferences Server requests exchanging the Authorization Code for an
+      Access Token and ID Token. This includes the following:
+      * `code`, the Authorization Code sent from the previous response from
+         Google
+      * `client_id` for the Preferences Server (same value as steps 2i)
+      * `redirect_uri` which is the endpoint on the Preferences Server that
+         will receive the response from Google (same value as step 2i)
+      * `grant_type` which must be set to `authorization_code`
+   7. Google responds at the previously specified redirect uri with the Access
+      Token and ID Token
+      * The `access_token` can be used by the Preferences Server to access the Google API
+      * The `id_token` is a [JWT](https://tools.ietf.org/html/rfc7519)
+        identifying the Resource Owner and is signed by Google.
+3. Authenticate and Authorize UIO
+   1. Using the `id_token` the Preferences Server authenticates the Resource
+      Owner
```

Comment:
We talked about this recently, but adding here for reference. For services such as GitHub that do not support OpenID Connect, there will be no `id_token` returned. It's handy to use the `id_token` if available, but we can use the `access_token` if not. However, we may need to make additional requests to the SSO to verify it's still valid.

Reply:
Yes, we discussed this at today's meeting, and decided not to use either.  The Preferences Server will generate a session key after it receives the access token from the SSO provider.  This is more secure since it keeps the values separate -- the access token is a "secret" shared between the Preferences Server and Google.  The Preferences Server will persist this session key, and pass it back to the Edge Proxy and UIO.  They, in turn, will use it to make requests of the Preferences Server.  I've added the changes to the document that explain this approach.

> However, we may need to make additional requests to the SSO to verify it's still valid.

We _will_ have to make additional requests since access tokens have an expiry and need to be refreshed.  We decided at the meeting to tie the session key's expiration to the access token's expiration and refresh the session key at the same time.